plan/physical_plan_builder.go

```diff
@@ -91,7 +91,25 @@ func (p *baseLogicalPlan) findBestTask(prop *requiredProp) (bestTask task, err e
 
 	bestTask = invalidTask
 	childTasks := make([]task, 0, len(p.children))
-	for _, pp := range p.self.exhaustPhysicalPlans(prop) {
+
+	// If prop.enforced is true, cols of prop as parameter in exhaustPhysicalPlans should be nil
+	// And reset it for enforcing task prop and storing map<prop,task>
+	oldPropCols := prop.cols
+	if prop.enforced {
+		// First, get the bestTask without enforced prop
+		prop.enforced = false
+		bestTask, err = p.findBestTask(prop)
```

Comment:
Maybe it's better to have a comparing with enforced task?

Reply:
There's no comparing in `DataSource.findBestTask`. It feels like no enforced task is always better than enforced task in `DataSource.findBestTask`.

Reply:
I think that may not be true. For example, for `select /*+ TIDB_SMJ(t1,t2)*/ * from t t1, t t2 where t1.a = t2.a order by t1.a desc;`, let's say there is an index on t.a, currently the result is
```
+----------------+--------------+-------------------------------+------+-------------------------------------------------------------------+----------+
| id             | parents      | children                      | task | operator info                                                     | count    |
+----------------+--------------+-------------------------------+------+-------------------------------------------------------------------+----------+
| IndexScan_12   |              |                               | cop  | table:t1, index:a, range:[<nil>,+inf], keep order:true            | 10000.00 |
| TableScan_13   |              |                               | cop  | table:t, keep order:false                                         | 10000.00 |
| IndexLookUp_14 | MergeJoin_10 |                               | root | index:IndexScan_12, table:TableScan_13                            | 10000.00 |
| IndexScan_16   |              |                               | cop  | table:t2, index:a, range:[<nil>,+inf], keep order:true            | 10000.00 |
| TableScan_17   |              |                               | cop  | table:t, keep order:false                                         | 10000.00 |
| IndexLookUp_18 | MergeJoin_10 |                               | root | index:IndexScan_16, table:TableScan_17                            | 10000.00 |
| MergeJoin_10   | Projection_9 | IndexLookUp_14,IndexLookUp_18 | root | inner join, equal:[eq(t1.a, t2.a)], left key:t1.a, right key:t2.a | 12500.00 |
| Projection_9   | Sort_7       | MergeJoin_10                  | root | t1.a, t1.b, t2.a, t2.b                                            | 12500.00 |
| Sort_7         |              | Projection_9                  | root | t1.a:desc                                                         | 12500.00 |
+----------------+--------------+-------------------------------+------+-------------------------------------------------------------------+----------+
```
But may be it is better to sort near data source.

Reply:
And, why do we sort at the end?

Reply:
The current logic is this: If there is an index in the join condition, then SMJ will not generate an enforce task to compare. This pr's propose is to fix "SMJ hint doesn't work" bug. With or without index is considered separately.

Reply:
Oh, you are right. Let's take a look at this example:
```
MySQL [test]> explain select * from t order by a desc;
+---------------+---------+---------------+------+----------------------------------------------+----------+
| id            | parents | children      | task | operator info                                | count    |
+---------------+---------+---------------+------+----------------------------------------------+----------+
| TableScan_7   |         |               | cop  | table:t, range:[-inf,+inf], keep order:false | 10000.00 |
| TableReader_8 | Sort_4  |               | root | data:TableScan_7                             | 10000.00 |
| Sort_4        |         | TableReader_8 | root | test.t.a:desc                                | 10000.00 |
+---------------+---------+---------------+------+----------------------------------------------+----------+
```
Even if there is an index on a, we choose to sort after, because the desc scan is quite slow.